buffer/src/main/java/io/netty/buffer/api/BufferAccessors.java

```diff
@@ -0,0 +1,614 @@
+     *                                   greater than {@link Buffer#capacity()} minus {@link Byte#BYTES}.
+     */
+    Buffer setUnsignedByte(int woff, int value);
+
+    /**
+     * Read the char value at the current {@link Buffer#readerOffset()},
+     * and increases the reader offset by 2.
+     * The value is read using a 2-byte UTF-16 encoding,
+     * with the {@link Buffer#order() configured} default byte order.
+     *
+     * @return The char value at the current reader offset.
+     * @throws IndexOutOfBoundsException If {@link Buffer#readableBytes} is less than 2.
+     */
+    char readChar();
+
+    /**
+     * Get the char value at the given reader offset.
+     * The {@link Buffer#readerOffset()} is not modified.
+     * The value is read using a 2-byte UTF-16 encoding,
+     * with the {@link Buffer#order() configured} default byte order.
+     *
+     * @param roff The read offset, an absolute offset into this buffer, to read from.
+     * @return The char value at the given offset.
+     * @throws IndexOutOfBoundsException if the given offset is out of bounds of the buffer, that is, less than 0 or
+     *                                   greater than {@link Buffer#capacity()} minus 2.
+     */
+    char getChar(int roff);
+
+    /**
+     * Write the given char value at the current {@link Buffer#writerOffset()},
+     * and increase the writer offset by 2.
+     * The value is written using a 2-byte UTF-16 encoding,
+     * with the {@link Buffer#order() configured} default byte order.
+     *
+     * @param value The char value to write.
+     * @return This Buffer.
+     * @throws IndexOutOfBoundsException If {@link Buffer#writableBytes} is less than 2.
+     */
+    Buffer writeChar(char value);
+
+    /**
+     * Set the given char value at the given write offset. The {@link Buffer#writerOffset()} is not modified.
+     * The value is written using a 2-byte UTF-16 encoding,
+     * with the {@link Buffer#order() configured} default byte order.
+     *
+     * @param woff The write offset, an absolute offset into this buffer to write to.
+     * @param value The char value to write.
+     * @return This Buffer.
+     * @throws IndexOutOfBoundsException if the given offset is out of bounds of the buffer, that is, less than 0 or
+     *                                   greater than {@link Buffer#capacity()} minus 2.
+     */
+    Buffer setChar(int woff, char value);
+
+    /**
+     * Read the short value at the current {@link Buffer#readerOffset()},
+     * and increases the reader offset by {@link Short#BYTES}.
+     * The value is read using a two's complement 16-bit encoding,
+     * with the {@link Buffer#order() configured} default byte order.
+     *
+     * @return The short value at the current reader offset.
+     * @throws IndexOutOfBoundsException If {@link Buffer#readableBytes} is less than {@link Short#BYTES}.
+     */
+    short readShort();
```

Comment:
Not sure where to put this comment, but it is *really* annoying in the Netty 4 byte buf API to implement a buffer due to all these specialized word accessors.   It would be helpful for new implementations to inherit some (albeit less efficient) default methods.   For example:

```java
default short readShort() {
  byte b1 = readByte();
  byte b2 = readByte();
  // endian handling left out for brevity.
  return b1 << 8 | b2;
}
```

And, arguing against myself here a little, it may look risky to call methods that have a slower default implementation (like InputStream.read()), but I think this is worth it.

Reply:
How often would you need to implement your own buffer, though?

Reply:
Pretty often I feel like.  (Most recently with Panama and jemalloc: https://github.com/netty/netty/pull/10335 ).   In most cases, the buffer I want only has one accessed method on it (like readint) so the other methods aren't ever going to be called.  In other cases, I want to pass around a non-null buffer for testing, but dont ever end up calling any methods on it.   I avoid mocking in general, so being able to make a small instance that just throws for most of the methods is what I want.

Reply:
So there are two levels of implementations, it sounds like. One is for testing, where you want a normal buffer, except one or two methods have been made to behave differently, for the purpose of the test. The other are serious implementations that would stand as alternatives to the built-in ones in production use.
For the former you want a delegating stub. For the latter, you're signing up for a lot of complicated work, and if you want the best performance you'll have to implement every accessor anyway, so I don't see much value in default methods for that use case.

Reply:
Added a `BufferStub`